### PR TITLE
ATM2018Q2 - Hometask M9

### PR DESCRIPTION
**Pull request by yauhen07**
Draft. Singleton and builer patterns

### CODE REVIEW

File: src/main/java/core/WebDriverSingleton.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        driver.manage().window().maximize();
+        return driver;
+    }
+
+    public static void kill() {
+        if (instance != null) {
+            try {
+                instance.quit();
+            } catch (Exception e) {
+                System.out.println("Cannot kill browser");
```
Comment (evgeny-tumashchick) [inline src/main/java/core/WebDriverSingleton.java]: all exceptions should be logged. In your case it is at least System.out.println(e.getMessage);

File: src/main/java/core/LocalCreator.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+public class LocalCreator extends RunLocationCreator {
+    protected static String PROPERTY_CHROME = "webdriver.chrome.driver";
+    protected static String PATH_TO_CHROME_DRIVER = "C:\\data\\tools\\chromedriver.exe";
+
+    protected WebDriver getWebDriver() {
+        System.setProperty(PROPERTY_CHROME, PATH_TO_CHROME_DRIVER);
+        driver = new ChromeDriver();
+        return driver;
+    }
+
```
Comment (evgeny-tumashchick) [inline src/main/java/core/LocalCreator.java]: remove empty line

File: src/test/java/com/epam/atm/selenium/LoginActionTest.java [hunk 3]
```diff
@@ -6,12 +6,12 @@
 import pageobject.HomePage;
 import pageobject.LoginPage;
 
-public class LoginActionTest extends BaseRemoteTest {
+public class LoginActionTest extends BaseTest {
```
Comment (evgeny-tumashchick) [inline src/test/java/com/epam/atm/selenium/LoginActionTest.java:9]: remove duplicate test

File: src/main/java/core/LocalCreator.java (hunk 2 above)
Comment (yauhen07) [reply src/main/java/core/LocalCreator.java]: done

File: src/main/java/core/WebDriverSingleton.java (hunk 1 above)
Comment (yauhen07) [reply src/main/java/core/WebDriverSingleton.java]: done

File: .gitignore [hunk 4]
```diff
@@ -3,4 +3,6 @@
 target
 logs
 .xml
-gherkin
\ No newline at end of file
+gherkin
+screenshots
+*.json
```
Comment (evgeny-tumashchick) [inline .gitignore]: json is a usual storage for project properties, better ignore selected file or package if u not need this one json file

File: src/main/java/core/CustomWebDriver.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+import org.openqa.selenium.WebElement;
+
+import java.util.List;
+import java.util.Set;
+
+public class CustomWebDriver implements WebDriver {
+    protected WebDriver driver;
+
+    public CustomWebDriver(WebDriver driver) {
+        this.driver = driver;
```
Comment (evgeny-tumashchick) [inline src/main/java/core/CustomWebDriver.java]: Just for better login lets add in constructor smth like
System.out.println("Custom web driver is used");

File: src/test/java/com/epam/atm/selenium/BaseTest.java [hunk 6]
```diff
@@ -1,36 +1,23 @@
...
-        System.out.println("Add Implicit wait");
+        runLocationCreator = new LocalCreator();
+        WebDriverSingleton.selectDriverType(runLocationCreator);
     }
 
     @AfterClass(description = "Close browser")
     public void closeBrowser() {
-        driver.quit();
+        WebDriverSingleton.kill();
         System.out.println("Browser was closed");
```
Comment (evgeny-tumashchick) [inline src/test/java/com/epam/atm/selenium/BaseTest.java]: now u have wrapper for web driver and can move this messages to Custom web driver

Comment (evgeny-tumashchick) [reply src/test/java/com/epam/atm/selenium/BaseTest.java]: Please look through other logging and move all that could be moved

File: src/test/java/com/epam/atm/selenium/LoginActionTest.java (hunk 3 above)
Comment (yauhen07) [reply src/test/java/com/epam/atm/selenium/LoginActionTest.java:9]: Leave it as is.

File: .gitignore (hunk 4 above)
Comment (yauhen07) [reply .gitignore]: Done

File: src/main/java/core/CustomWebDriver.java (hunk 5 above)
Comment (yauhen07) [reply src/main/java/core/CustomWebDriver.java]: Done

File: src/test/java/com/epam/atm/selenium/BaseTest.java (hunk 6 above)
Comment (yauhen07) [reply src/test/java/com/epam/atm/selenium/BaseTest.java]: Done

**Pull request merged**